### ISSUE TITLE
No longer compatible with Keil uVision coverage export

### ISSUE THREAD

**Issue by tacooper** [gcovr/gcovr#621]
According to the latest Keil MDK/uVision docs:
https://developer.arm.com/documentation/101407/0537/Debug-Commands/COVERAGE/COVERAGE-GCOV-Export?lang=en
`*.gcov` files are still generated with 6 octothorpes (`######`) instead of 5 (`#####`). 

This used to work before adding the new `gcovr/gcov_parser.py` in https://github.com/gcovr/gcovr/commit/304518453d352b21852e94f79ff78a4f4e4d2cb9.
Old parsing for any number of `#` symbols:
https://github.com/gcovr/gcovr/blob/aa7c0a1179fee66f7abc5b83586d4ec933844ed0/gcovr/gcov.py#L355
New parsing for only 5 `#` symbols:
https://github.com/gcovr/gcovr/blob/7279c52e1c33b00b593415cb967dbf1c5b2fc805/gcovr/gcov_parser.py#L768

Is it possible to update this parsing to account for lines with any number of `#` symbols? Or should other coverage tools (such as Keil uVision) be expected to update their coverage generation to use exactly 5 `#` symbols? Is this change or incompatibility documented anywhere?

**Desktop Environment:**
 - OS: Windows 10
 - Keil uVision: 5.37
 - Gcovr: 5.1
 - Python: 3.9.12

**Comment by Spacetown** [comment]
Thanks for reporting this.
In our changelig there is following point (in internal changes):
```
New parser for .gcov file format, should be more robust. (#512)
```
We expected less problems but now this is the second issue. The first issue which was detected by the new parser seems to be a memory corruption in the coverage counters which produces negative values for GCC.

In the documentation of GCC for invoking gcov (https://gcc.gnu.org/onlinedocs/gcc-11.1.0/gcc/Invoking-Gcov.html) the format is described as follow:
```
Unexecuted lines are marked ‘#####’ or ‘=====’, depending on whether they are reachable by non-exceptional paths or only exceptional paths such as C++ exception handlers, respectively. Given the ‘-a’ option, unexecuted blocks are marked ‘$$$$$’ or ‘%%%%%’, depending on whether a basic block is reachable via non-exceptional or exceptional paths.
```

According to this, the previous behavior was wrong. Please raise a bug report for the Keil MDK/uVision.

**Comment by tacooper** [comment]
Thanks for your quick and helpful attention. I have opened a Keil MDK support case.

**Comment by Spacetown** [comment]
Please can you place a link here?

**Comment by tacooper** [comment]
> 

Yes, sorry I wasn't sure if it would be accessible because it seems like an ARM account is required to create/see the case:
https://services.arm.com/support/s/case/5004L00000IdvaUQAR/coverage-export-is-incompatible-with-standard-gccgcov-and-thus-latest-gcovr
Here it is though, in case it is viewable by those with ARM accounts (I can't verify).

**Comment by Spacetown** [comment]
Yes, you need an account. But than it should be visible

**Comment by tacooper** [comment]
It appears that the issue was forwarded internally (as indicated by the case's Emails tab) and my case was closed. I received the following separate email:

> We would like to inform you that a product Defect/Enhancement has been logged on your behalf for your Support Case [00448557 - Coverage export is incompatible with standard GCC/Gcov (and thus latest Gcovr)](https://services.arm.com/support/s/case/5004L00000IdvaU). There is no further action that you need to take at this time. We will aim to keep you informed of any changes to the defect/enhancement with periodic updates. You will continue to be able to view the case at any time in the My Defects/Enhancements tab within the [Support Hub](https://services.arm.com/support/s/case/5004L00000IdvaU).

I guess it falls to me to keep any status/progress up-to-date on this ticket... 🙃

**Comment by chentang16** [comment]
As written in the Email sent from us (Arm) "... We will aim to keep you informed of any changes to the defect/enhancement with periodic update..."
Since the uVision debugger regarding the change of exporting coverage info isn't open source like gcovr here in Github, unfortunately the internal discussion/fix/change about this specific bug-fix is not visible to you

**Comment by tacooper** [comment]
> Since the uVision debugger regarding the change of exporting coverage info isn't open source like gcovr here in Github, unfortunately the internal discussion/fix/change about this specific bug-fix is not visible to you

Understood. I only meant that since the support case was closed and may no longer be updated in the future, I would personally see fit to keep this ticket manually updated with anything I hear back.

**Comment by latk** [comment]
I created a patch that should fix this issue (#623) but am not sure if it should be merged. One one hand, supporting 6-character markers in addition to 5-character markers is an easy fix. On the other hand, it's going to be tech debt in favour of a third party tool that gcovr cannot reasonably support. If ARM is going to eventually fix their tool I'd rather not merge my fix, but of course users could apply it manually to their gcovr installation.

Regarding the robustness of the new parser: it is more robust in a sense that it more obviously implements exactly the data format provided by GCC gcov. If there is a deviation, the new parser will loudly complain and should then be able to recover by continuing to the next source code line. The old parser accepted more input files, but at the risk of interpreting those files incorrectly.
